src/java/org/apache/cassandra/cql3/selection/Selectable.java

```diff
@@ -245,18 +271,20 @@ public Selector.Factory newSelectorFactory(TableMetadata table,
             if (column.isPrimaryKeyColumn())
                 throw new InvalidRequestException(
                         String.format("Cannot use selection function %s on PRIMARY KEY part %s",
-                                      isWritetime ? "writeTime" : "ttl",
+                                      kind.name,
                                       column.name));
-            if (column.type.isCollection())
+
+            // maxwritetime is allowed for collection
+            if (column.type.isCollection() && !kind.allowedForMultiCell())
```

Comment:
Yes and nice catch. It should use `isMultiCell` instead of `isCollection`. Otherwise, `writetime` function on non-frozen UDT is allowed. I was planning to fix it later, since it is relatively unrelated with the patch.

Reply:
I guess we can fix it in a separate bugfix ticket, also targeting past releases.

Reply:
Just opened [CASSANDRA-17628](https://issues.apache.org/jira/browse/CASSANDRA-17628) to deal with this.